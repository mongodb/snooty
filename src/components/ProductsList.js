import React, { useState } from 'react';
import styled from '@emotion/styled';
import Icon from '@leafygreen-ui/icon';
import { uiColors } from '@leafygreen-ui/palette';
import Link from './Link';
import { useAllProducts } from '../hooks/useAllProducts';
import { theme } from '../theme/docsTheme';

const HeadingTitle = styled('span')`
  padding-left: ${theme.size.small};
`;

const Products = styled(`ul`)`
  list-style-type: none;
  padding: 0;
  ${({ isOpen }) => !isOpen && 'display: none;'}
`;

const ProductsListContainer = styled('div')`
  line-height: 20px;
  width: 100%;
`;

const ProductsListHeading = styled('div')`
  align-items: center;
  color: ${({ isOpen }) => (isOpen ? uiColors.gray.dark3 : uiColors.gray.dark1)};
  cursor: pointer;
  display: flex;
  padding: 0px ${theme.size.medium} 12px;
  user-select: none;

  :hover {
    color: ${uiColors.gray.dark3};
  }
`;

const ProductLink = styled(Link)`
  color: ${uiColors.gray.dark2};
  display: inline-block;
  font-size: ${theme.fontSize.small};
  letter-spacing: 0;
  padding: ${theme.size.tiny} ${theme.size.medium} ${theme.size.tiny} 44px;
  width: 100%;

  :hover {
    color: ${uiColors.gray.dark2};
    font-weight: bold;
    text-decoration: none;
  }

  @media ${theme.screenSize.upToSmall} {
    padding: ${theme.size.small} ${theme.size.medium} ${theme.size.small} 44px;
  }
`;

const StyledIcon = styled(Icon)`
  height: 12px;
  width: 12px;
`;

const ProductsList = () => {
  const products = useAllProducts();
  const [isOpen, setOpen] = useState(false);

  return (
    <ProductsListContainer>
      <ProductsListHeading isOpen={isOpen} onClick={() => setOpen(!isOpen)}>
        <StyledIcon glyph={isOpen ? 'ChevronUp' : 'ChevronDown'} />
        <HeadingTitle>View all products</HeadingTitle>
      </ProductsListHeading>
      <Products isOpen={isOpen}>
        {products.map(({ title, url }, index) => {
          return (
            <li key={index}>
              <ProductLink to={url}>{title}</ProductLink>
            </li>
          );
        })}
      </Products>
    </ProductsListContainer>
  );
};

export default ProductsList;
