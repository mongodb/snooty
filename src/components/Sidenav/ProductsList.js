import React, { useState } from 'react';
import { CSSTransition } from 'react-transition-group';
import { Global, css } from '@emotion/core';
import styled from '@emotion/styled';
import { css as LeafyCSS, cx } from '@leafygreen-ui/emotion';
import Icon from '@leafygreen-ui/icon';
import { uiColors } from '@leafygreen-ui/palette';
import Link from '../Link';
import { useAllProducts } from '../../hooks/useAllProducts';
import { theme } from '../../theme/docsTheme';

const chevronRotationDuration = '200ms';

const openTransition = css`
  transition: all 300ms ease-in-out ${chevronRotationDuration};
`;

const transitionClasses = css`
  .products-list-enter {
    display: block !important;
    margin-top: -100vh;
  }
  .products-list-enter-active {
    display: block !important;
    margin-top: 0;
    ${openTransition}
  }
  .products-list-enter-done {
    display: block !important;
  }
  .products-list-exit {
    display: block !important;
    margin-top: 0;
  }
  .products-list-exit-active {
    display: block !important;
    margin-top: -100vh;
    ${openTransition}
  }
`;

const HeadingTitle = styled('span')`
  padding-left: ${theme.size.small};
`;

const Products = styled(`ul`)`
  display: none;
  list-style-type: none;
  padding: 0;
`;

const ProductsListContainer = styled('div')`
  line-height: 20px;
  width: 100%;
`;

const ProductsListHeading = styled('div')`
  align-items: center;
  background-color: ${uiColors.gray.light3};
  color: ${({ isOpen }) => (isOpen ? uiColors.gray.dark3 : uiColors.gray.dark1)};
  cursor: pointer;
  display: flex;
  padding: ${theme.size.default} ${theme.size.medium} 12px;
  position: relative;
  user-select: none;
  z-index: 1;

  :hover {
    color: ${uiColors.gray.dark3};
  }
`;

const ProductLink = styled(Link)`
  color: ${uiColors.gray.dark2};
  display: inline-block;
  font-size: ${theme.fontSize.small};
  letter-spacing: 0;
  padding: ${theme.size.tiny} ${theme.size.medium} ${theme.size.tiny} 44px;
  width: 100%;

  :hover {
    color: ${uiColors.gray.dark2};
    font-weight: bold;
    text-decoration: none;
  }

  @media ${theme.screenSize.upToSmall} {
    padding: ${theme.size.small} ${theme.size.medium} ${theme.size.small} 44px;
  }
`;

const iconStyle = ({ isOpen }) => LeafyCSS`
  height: 12px;
  width: 12px;

  transform: rotate(${isOpen ? 0 : -180}deg);
  transition: transform ${chevronRotationDuration};
`;

const ProductsList = () => {
  const products = useAllProducts();
  const [isOpen, setOpen] = useState(false);

  return (
    <>
      <Global styles={transitionClasses} />
      <ProductsListContainer>
        <ProductsListHeading isOpen={isOpen} onClick={() => setOpen(!isOpen)}>
          <Icon className={cx(iconStyle({ isOpen }))} glyph="ChevronUp" />
          <HeadingTitle>View all products</HeadingTitle>
        </ProductsListHeading>
      </ProductsListContainer>
      <CSSTransition
        in={isOpen}
        classNames="products-list"
        addEndListener={(node, done) => node.addEventListener('transitionend', done, false)}
      >
        <Products>
          {products.map(({ title, url }, index) => {
            return (
              <li key={index}>
                <ProductLink to={url}>{title}</ProductLink>
              </li>
            );
          })}
        </Products>
      </CSSTransition>
    </>
  );
};

export default ProductsList;
