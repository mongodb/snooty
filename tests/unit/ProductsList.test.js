import React from 'react';
import { mount } from 'enzyme';
import { matchers } from 'jest-emotion';
import ProductsList from '../../src/components/ProductsList';

const mockProducts = [
  {
    title: 'MongoDB Drivers',
    url: 'https://www.docs.mongodb.com/drivers/',
  },
  {
    title: 'MongoDB Atlas',
    url: 'https://www.docs.atlas.mongodb.com/',
  },
];

jest.mock('../../src/hooks/useAllProducts', () => ({
  useAllProducts: () => mockProducts,
}));

expect.extend(matchers);

describe('ProductsList', () => {
  it('renders with products', async () => {
    let wrapper = mount(<ProductsList />);
    expect(wrapper).toMatchSnapshot();

    // Display products list
    wrapper.find('ProductsListHeading').simulate('click');
    expect(wrapper.find('Products').props()).toHaveProperty('isOpen', true);
    expect(wrapper.find('Products')).not.toHaveStyleRule('display', 'none');
  });
});
