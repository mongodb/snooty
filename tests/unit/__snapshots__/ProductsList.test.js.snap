// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductsList renders with products 1`] = `
.emotion-22 {
  line-height: 20px;
  width: 100%;
}

.emotion-10 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #5D6C74;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0px 24px 12px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.emotion-10:hover {
  color: #21313C;
}

.emotion-0 {
  height: 12px;
  width: 12px;
}

.emotion-8 {
  padding-left: 8px;
}

.emotion-20 {
  list-style-type: none;
  padding: 0;
  display: none;
}

.emotion-12 {
  color: #3D4F58;
  display: inline-block;
  font-size: 14px;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  padding: 4px 24px 4px 44px;
  width: 100%;
}

.emotion-12:hover {
  color: #3D4F58;
  font-weight: bold;
  -webkit-text-decoration: none;
  text-decoration: none;
}

@media only screen and (max-width:420px) {
  .emotion-12 {
    padding: 8px 24px 8px 44px;
  }
}

<ProductsList>
  <ProductsListContainer>
    <div
      className="emotion-22 emotion-23"
    >
      <ProductsListHeading
        isOpen={false}
        onClick={[Function]}
      >
        <div
          className="emotion-10 emotion-11"
          onClick={[Function]}
        >
          <StyledIcon
            glyph="ChevronDown"
          >
            <Icon
              className="emotion-0 emotion-1"
              glyph="ChevronDown"
            >
              <ChevronDown
                className="emotion-0 emotion-1"
              >
                <ChevronDown
                  aria-label="Chevron Down Icon"
                  className="emotion-0 emotion-1"
                  height={16}
                  role="img"
                  width={16}
                >
                  <svg
                    aria-label="Chevron Down Icon"
                    className="emotion-0 emotion-1"
                    fill="none"
                    height={16}
                    role="img"
                    viewBox="0 0 16 16"
                    width={16}
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      clipRule="evenodd"
                      d="M1.636 5.364a1 1 0 000 1.414l4.95 4.95.707.707a1 1 0 001.414 0l.707-.707 4.95-4.95a1 1 0 000-1.414l-.707-.707a1 1 0 00-1.414 0L8 8.899 3.757 4.657a1 1 0 00-1.414 0l-.707.707z"
                      fill="currentColor"
                      fillRule="evenodd"
                    />
                  </svg>
                </ChevronDown>
              </ChevronDown>
            </Icon>
          </StyledIcon>
          <HeadingTitle>
            <span
              className="emotion-8 emotion-9"
            >
              View all products
            </span>
          </HeadingTitle>
        </div>
      </ProductsListHeading>
      <Products
        isOpen={false}
      >
        <ul
          className="emotion-20 emotion-21"
        >
          <li
            key="0"
          >
            <ProductLink
              to="https://www.docs.mongodb.com/drivers/"
            >
              <Link
                className="emotion-12 emotion-13"
                to="https://www.docs.mongodb.com/drivers/"
              >
                <a
                  className="emotion-12 emotion-13"
                  href="https://www.docs.mongodb.com/drivers/"
                >
                  MongoDB Drivers
                </a>
              </Link>
            </ProductLink>
          </li>
          <li
            key="1"
          >
            <ProductLink
              to="https://www.docs.atlas.mongodb.com/"
            >
              <Link
                className="emotion-12 emotion-13"
                to="https://www.docs.atlas.mongodb.com/"
              >
                <a
                  className="emotion-12 emotion-13"
                  href="https://www.docs.atlas.mongodb.com/"
                >
                  MongoDB Atlas
                </a>
              </Link>
            </ProductLink>
          </li>
        </ul>
      </Products>
    </div>
  </ProductsListContainer>
</ProductsList>
`;
